packages/gatsby/src/schema/resolvers.ts

```diff
@@ -221,20 +224,42 @@ export function link<TSource, TArgs>(
     TArgs
   >
 ): GatsbyResolver<TSource, TArgs> {
-  return async function linkResolver(
+  // Note: we explicitly make an attempt to prevent using the `async` keyword because often
+  //       it does not return a promise and this makes a significant difference at scale.
+
+  return function linkResolver(
     source,
     args,
     context,
     info
-  ): Promise<any> {
+  ): ResolvedLink | Promise<ResolvedLink> {
     const resolver = fieldConfig.resolve || context.defaultFieldResolver
-    const fieldValue = await resolver(source, args, context, {
+    const fieldValueOrPromise = resolver(source, args, context, {
       ...info,
       from: options.from || info.from,
       fromNode: options.from ? options.fromNode : info.fromNode,
     })
 
-    if (fieldValue == null) return null
+    // Note: for this function, at scale, conditional .then is more efficient than generic await
+    if (fieldValueOrPromise?.then) {
+      return fieldValueOrPromise.then(
+        fieldValue => linkResolverValue(fieldValue, args, context, info),
+        err => Promise.reject(err)
```

Comment:
The reject handler seems redundant. I think Promise rejection is propagated anyway?

```suggestion
        fieldValue => linkResolverValue(fieldValue, args, context, info)
```